nifi-extension-bundles/nifi-protobuf-bundle/nifi-protobuf-services/src/main/java/org/apache/nifi/services/protobuf/ProtobufSchemaCompiler.java

```diff
@@ -0,0 +1,245 @@
+            safeDeleteDirectory(tempDir);
+        }
+    }
+
+    private Schema createAndLoadSchema(final Path tempDir) {
+        final SchemaLoader schemaLoader = new SchemaLoader(FileSystems.getDefault());
+
+        final List<Location> roots = new ArrayList<>();
+        roots.add(Location.get(tempDir.toString()));
+
+        // Add standard protobuf libraries
+        roots.addAll(STANDARD_PROTOBUF_LOCATIONS);
+
+        schemaLoader.initRoots(roots, Collections.emptyList());
+
+        // Load and return the compiled schema
+        return schemaLoader.loadSchema();
+    }
+
+    private void safeDeleteDirectory(final Path directory) {
+        if (Files.exists(directory)) {
+            try {
+                FileUtils.deleteDirectory(directory.toFile());
+            } catch (final IOException | IllegalArgumentException e) {
+                logger.warn("Failed to delete temporary directory: {}", directory, e);
+            }
+        }
+    }
+
+
+    /**
+     * Writes a schema definition to the temporary directory structure.
+     * If package name is present, creates the appropriate directory structure.
+     *
+     * @param tempDir          the temporary directory root
+     * @param schemaDefinition the schema definition to write
+     * @throws IOException if unable to create directories or write files
+     */
+    private void writeSchemaToTempDirectory(final Path tempDir, final SchemaDefinition schemaDefinition) throws IOException {
+        logger.debug("Writing schema definition to temporary directory. Identifier: {}", schemaDefinition.getIdentifier());
+
+        final String schemaFileName = generateSchemaFileName(schemaDefinition);
+        final Path schemaFile = tempDir.resolve(schemaFileName);
+
+        // Write schema text to file
+        Files.write(schemaFile, schemaDefinition.getText().getBytes(), CREATE, WRITE, TRUNCATE_EXISTING);
+        logger.debug("Successfully wrote schema to file: {} (string length: {})",
+            schemaFile, schemaDefinition.getText().length());
+    }
+
+    /**
+     * Generates a filename for a schema definition, ensuring it has a .proto extension.
+     *
+     * @param schemaDefinition the schema definition
+     * @return the generated filename
+     */
+    private String generateSchemaFileName(final SchemaDefinition schemaDefinition) {
+        String schemaFileName = schemaDefinition.getIdentifier().getName().orElseGet(
+            () -> String.valueOf(schemaDefinition.getIdentifier().getSchemaVersionId().orElse(0L))
+        );
+
+        if (!schemaFileName.endsWith(".proto")) {
+            schemaFileName += ".proto"; // Ensure the file ends with .proto, otherwise the wire library will not recognize it
+        }
+
+        return schemaFileName;
+    }
+
+    private void processSchemaReferences(final Path tempDir, final Map<String, SchemaDefinition> references) throws IOException {
+        logger.debug("Processing schema references. Count: {}, Temp directory: {}",
```

Comment:
Recommend adjusting the wording:
```suggestion
        logger.debug("Processing [{}] schema references in [{}]",
```